### PR TITLE
Adding in wrapper for Drupal node redirect and business logic to support

### PR DESCRIPTION
**Pull request by Dillon-Dumesnil**
when to add a redirect.

### CODE REVIEW

File: course_discovery/apps/course_metadata/tests/test_publishers.py [hunk 1]
```diff
@@ -349,6 +349,46 @@ def test_alias(self):
...
+        self.mock_node_retrieval(self.publisher.node_lookup_field, lookup_value)
+        lookup_value = getattr(previous_obj, self.publisher.unique_field)
+        self.mock_node_retrieval(self.publisher.node_lookup_field, lookup_value)
+
+        self.mock_get_redirect_form()
+        self.mock_add_redirect(self.obj, previous_obj)
+
+        self.publisher.add_url_redirect(self.obj, previous_obj)
+
+        assert responses.calls[-1].request.url == '{}/add'.format(self.publisher.redirect_api_base)
```
Comment (mikix) [inline course_discovery/apps/course_metadata/tests/test_publishers.py]: Good line. I wish `responses` had a method you could call like `responses.assertAllRegisteredCallsActivated()` or something.  But in absence of that, good to check manually. It *does* check that itself when used as a context manager. So an alternative would be to restructure the test that way. But this is fine.

I find it so odd that some of our test classes don't inherit from `TestCase` and thus don't have `self.assertEqual`.

File: course_discovery/apps/course_metadata/tests/test_publishers.py [hunk 2]
```diff
@@ -349,6 +349,46 @@ def test_alias(self):
...
+
+        responses.reset()
+
+        # Same scenario, but this time a non-200 status code is returned during
+        # redirect creation. An exception should be raised.
+        self.mock_api_client()
```
Comment (mikix) [inline course_discovery/apps/course_metadata/tests/test_publishers.py]: optional nit: This second block is probably more logically a second test method. Or use ddt.

File: course_discovery/apps/publisher/models.py [hunk 3]
```diff
@@ -995,7 +1001,31 @@ def approved(self):
...
+        # These will be unpublished in favor of the new course.
+        for run in discovery_course.course_runs.all():
+            if run != discovery_run and run.status == CourseRunStatus.Published and run.end and run.end < now:
+                CourseRunMarketingSitePublisher(site.partner).add_url_redirect(discovery_run, run)
+                run.status = CourseRunStatus.Unpublished
+                run.save()
+
+        # Notify course team
+        if waffle.switch_is_active('enable_publisher_email_notifications'):
+            emails.send_course_run_published_email(self.course_run, site)
```
Comment (mikix) [inline course_discovery/apps/publisher/models.py:1028]: Guh, this whole method sucks, man. #bebetter

File: course_discovery/apps/publisher/tests/test_models.py [hunk 4]
```diff
@@ -1106,7 +1110,7 @@ def test_internal_user_status(self):
 
 
 @ddt.ddt
-class CourseRunStateTests(TestCase):
+class CourseRunStateTests(MarketingSitePublisherTestMixin):
```
Comment (mikix) [inline course_discovery/apps/publisher/tests/test_models.py:1113]: Humph. I think this chain of mixins is abusing the word. Shouldn't take place of base class like that. Not your doing, and not worth changing. Just harrumphing.

**Review by mikix — APPROVED** [review]
Code looks good to me! We tested this together. Let's get this on stage and test further.

File: course_discovery/apps/publisher/models.py (hunk 3 above)
Comment (Dillon-Dumesnil) [reply course_discovery/apps/publisher/models.py:1028]: +1

**Pull request merged**